idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/KtAnalysisSession.kt

```diff
@@ -87,4 +91,9 @@ abstract class KtAnalysisSession(final override val token: ValidityToken) : Vali
 
     internal val expressionInfoProvider: KtExpressionInfoProvider get() = expressionInfoProviderImpl
     protected abstract val expressionInfoProviderImpl: KtExpressionInfoProvider
+
+    val KtType.isNullable: Boolean
```

Comment:
Is this really a part of `KtAnalysisSession`? Why not in `ktTypeUtils.kt`?

Reply:
I added it here so it has the `KtAnalysisSession` as a receiver. Moving it to ktTypeUtils would require explicitly passing the `KtAnalysisSession` as a parameter, which would be ugly at call-site. @darthorimar What do you think?

As for the difference between this and `isMarkedNullable`, `isMarkedNullable` is what it literally means. This does the proper type check instead. The difference between this two can be observed for type parameter `T:Any?`, which is not marked nullable but is nullable. Intuitively speaking, assuming types form a complete lattice (with intersection and union being meet and join and a explicit singleton type `Null`), if a type A `isMarkedNullable`, then we have `Null < A`. While if a type `isNullable`, then we have `A < (Null ∪ A)`

Reply:
call-site may use `with` to change the receiver (which may be ugly as well). I prefer explicit parameter, and leave "session" as a placeholder for those providers and any other -ers that do some tasks. But, it's just my opinion.

Regarding `isNullable` itself, there are many utils in `FirTypeUtils.kt`. I think `ConeKotlinType.canBeNull` will work as well (without potentially heavy subtype check). Since `KtFirType` has `coneType`, maybe `isNullable` can be abstract, and delegate to the implementer (`KtFirType`)? That way, we may not need to use `KtAnalysisSession` at all?

Reply:
Thanks! Delegate to `canBeNull` sounds good! I think with the current code structure, I can add a extension method for `KtType` inside `KtTypeProvider`.